src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.cs

```diff
@@ -61,30 +62,82 @@ public class FixAllContext
+        /// (a) False => Return only diagnostics with <see cref="Location.None"/>.
+        /// (b) True => Return all project diagnostics, regardless of whether or not they have a location.
+        /// </param>
+        /// <param name="cancellationToken">Cancellation token for fix all computation.</param>
+        public FixAllContext(
             Document document,
             CodeFixProvider codeFixProvider,
             FixAllScope scope,
-            string codeActionId,
+            string codeActionEquivalenceKey,
             IEnumerable<string> diagnosticIds,
             Func<Document, ImmutableHashSet<string>, CancellationToken, Task<IEnumerable<Diagnostic>>> getDocumentDiagnosticsAsync,
             Func<Project, bool, ImmutableHashSet<string>, CancellationToken, Task<IEnumerable<Diagnostic>>> getProjectDiagnosticsAsync,
             CancellationToken cancellationToken)
-            : this(document, document.Project, codeFixProvider, scope, codeActionId, diagnosticIds, getDocumentDiagnosticsAsync, getProjectDiagnosticsAsync, cancellationToken)
+            : this(document, document.Project, codeFixProvider, scope, codeActionEquivalenceKey, diagnosticIds, getDocumentDiagnosticsAsync, getProjectDiagnosticsAsync, cancellationToken)
         {
+            if (document == null)
+            {
+                throw new ArgumentNullException(nameof(document));
+            }
         }
 
-        internal FixAllContext(
+        /// <summary>
+        /// Creates a new <see cref="FixAllContext"/>.
+        /// Use this overload when applying fix all to a diagnostic with no source location, i.e. <see cref="Location.None"/>.
+        /// </summary>
+        /// <param name="project">Project within which fix all occurrences was triggered.</param>
+        /// <param name="codeFixProvider">Underlying <see cref="CodeFixes.CodeFixProvider"/> which triggered this fix all.</param>
+        /// <param name="scope"><see cref="FixAllScope"/> to fix all occurrences.</param>
+        /// <param name="codeActionEquivalenceKey">The <see cref="CodeAction.EquivalenceKey"/> value expected of a <see cref="CodeAction"/> participating in this fix all.</param>
+        /// <param name="diagnosticIds">Diagnostic Ids to fix.</param>
+        /// <param name="getDocumentDiagnosticsAsync">
+        /// Delegate to fetch diagnostics for any given document within the given fix all <paramref name="scope"/>.
+        /// This delegate is invoked by <see cref="GetDocumentDiagnosticsAsync(Document)"/> with the given <paramref name="diagnosticIds"/> and <paramref name="cancellationToken"/> as arguments.
+        /// </param>
+        /// <param name="getProjectDiagnosticsAsync">
+        /// Delegate to fetch diagnostics for any given project within the given fix all <paramref name="scope"/>.
+        /// This delegate is invoked by <see cref="GetProjectDiagnosticsAsync(Project)"/> and <see cref="GetAllDiagnosticsAsync(Project)"/>
```

Comment:
I am not fan of func in API, specially when it is complex. can it be just interface or an abstract class rather than two func? like AbstractDiagnosticProvider that has two methods

GetDocumentDiagnosticsAsync(Document, CancellationToken)
GetProjectDiagnosticsAsync(Project, CancellationToken). ImmutableHashSet<string> is implementation detail, and no reason to be exposed.